File: src/main/java/me/fourteendoggo/xkingdoms/storage/persistence/Storage.java
package me.fourteendoggo.xkingdoms.storage.persistence;

import me.fourteendoggo.xkingdoms.player.KingdomPlayer;
import org.jetbrains.annotations.Nullable;

import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.function.Supplier;
import java.util.logging.Level;
import java.util.logging.Logger;

public class Storage {
    private final PersistenceHandler persistenceHandler;
    private final Logger logger;

    public Storage(PersistenceHandler persistenceHandler, Logger logger) {
        this.persistenceHandler = persistenceHandler;
        this.logger = logger;
    }

    /* synchronous operations */

    public boolean connect() {
        try {
            persistenceHandler.connect();
            if (persistenceHandler instanceof Database database) {
                database.executePatches(logger);
            }
            return true;
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Failed to connect to the database", e);
        }
        return false;
    }

    public void disconnect() {
        try {
            persistenceHandler.disconnect();
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Failed to close database connection", e);
        }
    }

    @Nullable
    public KingdomPlayer loadPlayerSync(UUID id) {
        try {
            KingdomPlayer player = persistenceHandler.loadPlayer(id);
            logger.info("Loaded the player with uuid " + id);
            return player;
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Failed to load the player with uuid " + id, e);
        }
        return null;
    }

    public void savePlayerSync(KingdomPlayer player) {
        try {
            persistenceHandler.savePlayer(player);
            logger.info("Saved player " + player.getPlayer().getName());
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Failed to save player " + player.getPlayer().getName(), e);
        }
    }

    /* asynchronous operations */

    public CompletableFuture<KingdomPlayer> loadPlayer(UUID id) {
        return makeFuture(() -> persistenceHandler.loadPlayer(id),
                "Loaded the player with uuid " + id,
                "Failed to load the player with uuid " + id);
    }

    public CompletableFuture<Void> savePlayer(KingdomPlayer player) {
        return makeFuture(() -> persistenceHandler.savePlayer(player),
                "Saved player " + player.getPlayer().getName(),
                "Failed to save player " + player.getPlayer().getName());
    }

    /* when things go wrong, only messaging the player is required, exception is handled here */

    /**
     * Creates a {@link CompletableFuture} from the given supplier
     * The database implementation we wrap doesn't do exception handling, we persistenceHandler it to the completable future
     * So this will handle possible exceptions that will occur and put them in the future object
     *
     * @param supplier the action to perform
     * @param logOnSuccess the message that needs to be logged when this operation completed successfully
     * @param logOnFailure the message that needs to be logged if an exception occurred while trying to compute the result
     * @param <T> the return type of the inner operation, this is the same type which the future will return
     * @return a future which either holds the return type of the operation, or a {@link Throwable} object
     */
    private <T> CompletableFuture<T> makeFuture(Supplier<T> supplier, String logOnSuccess, String logOnFailure) {
        return CompletableFuture.supplyAsync(supplier).whenComplete((t, throwable) -> {
            if (throwable != null) {
                logger.log(Level.SEVERE, logOnFailure, throwable);
            } else {
                logger.info(logOnSuccess);
            }
        });
    }

    /**
     * @see Storage#makeFuture(Supplier, String, String) but this uses a runnable
     */
    private CompletableFuture<Void> makeFuture(Runnable runnable, String logOnSuccess, String logOnFailure) {
        return CompletableFuture.runAsync(runnable).whenComplete((v, throwable) -> {
            if (throwable != null) {
                logger.log(Level.SEVERE, logOnFailure, throwable);
            } else {
                logger.info(logOnSuccess);
            }
        });
    }
}

File: src/main/java/me/fourteendoggo/xkingdoms/storage/persistence/Storage.kt
package me.fourteendoggo.xkingdoms.storage.persistence

import me.fourteendoggo.xkingdoms.utils.Home
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.function.Supplier
import java.util.logging.Level
import java.util.logging.Logger

class Storage(private val persistenceHandler: PersistenceHandler, private val logger: Logger) {

    /* synchronous operations */
    fun connect(): Boolean {
        try {
            if (persistenceHandler is Database) {
                persistenceHandler.connect()
                return persistenceHandler.executePatches(logger)
            }
            return true
        } catch (e: Exception) {
            logger.log(Level.SEVERE, "Failed to connect to the database", e)
        }
        return false
    }

    fun disconnect() {
        try {
            if (persistenceHandler is Database) {
                persistenceHandler.disconnect()
            }
        } catch (e: Exception) {
            logger.log(Level.SEVERE, "Failed to close database connection", e)
        }
    }

    fun loadPlayerBlocking(id: UUID): KingdomPlayer? {
        try {
            val player = persistenceHandler.loadPlayer(id)
            logger.info("Loaded the player with uuid $id")
            return player
        } catch (e: Exception) {
            logger.log(Level.SEVERE, "Failed to load the player with uuid $id", e)
        }
        return null
    }

    fun savePlayerBlocking(player: KingdomPlayer) {
        try {
            persistenceHandler.savePlayer(player)
            logger.info("Loaded player " + player.player.name)
        } catch (e: Exception) {
            logger.log(Level.SEVERE, "Failed to load player " + player.player.name, e)
        }
    }

    /* asynchronous operations */
    fun loadPlayer(id: UUID): CompletableFuture<KingdomPlayer?> = makeFuture<KingdomPlayer?>(
            { persistenceHandler.loadPlayer(id) },
            "Loaded the player with uuid $id",
            "Failed to load the player with uuid $id")

    fun deleteHome(home: Home): CompletableFuture<Void> = makeFuture(
            { persistenceHandler.deleteHome(home) },
            "Deleted home " + home.name + " for " + home.owner,
            "Failed to delete home " + home.name + " for " + home.owner)

    fun savePlayer(player: KingdomPlayer): CompletableFuture<Void> = makeFuture(
            { persistenceHandler.savePlayer(player) },
            "Saved player " + player.player.name,
            "Failed to save player " + player.player.name)
    /* when things go wrong, only messaging the player is required, exception is handled here */
    /**
     * Creates a [CompletableFuture] from the given supplier
     * The database implementation we wrap doesn't do exception handling, we delegate it to the completable future
     * So this will handle possible exceptions that will occur and put them in the future object
     *
     * @param supplier     the action to perform
     * @param logOnSuccess the message that needs to be logged when this operation completed successfully
     * @param logOnFailure the message that needs to be logged if an exception occurred while trying to compute the result
     * @param <T>          the return type of the inner operation, this is the same type which the future will return
     * @return a future which either holds the return type of the operation, or a [Throwable] object
    </T> */
    private fun <T> makeFuture(supplier: Supplier<T>, logOnSuccess: String, logOnFailure: String): CompletableFuture<T> {
        return CompletableFuture.supplyAsync(supplier).whenComplete { t: T, throwable: Throwable? ->
            if (throwable != null) {
                logger.log(Level.SEVERE, logOnFailure, throwable)
            } else {
                logger.info(logOnSuccess)
            }
        }
    }

    /**
     * @see Storage.makeFuture
     */
    private fun makeFuture(runnable: Runnable, logOnSuccess: String, logOnFailure: String): CompletableFuture<Void> {
        return CompletableFuture.runAsync(runnable).whenComplete { v: Void?, throwable: Throwable? ->
            if (throwable != null) {
                logger.log(Level.SEVERE, logOnFailure, throwable)
            } else {
                logger.info(logOnSuccess)
            }
        }
    }
}
